### DISCUSSION TITLE
HS Example Issues

### DISCUSSION THREAD

**Discussion by BennyEvans** [hathach/tinyusb#955] [Q&A] [answered]
Hi,

I've been experimenting with the STM32H7 series and thought I'd give tinyUSB a go as STs USB library doesn't seem great. I've been using a Waveshare STM32H743 board (ensure you check for any solder bridges on these boards - I quickly found one on the LCD ribbon connector) with USB3300 PHY. This seems to be the same board discussed breifly in discussion #633 by @squintz and @MimitechIndustries

I can run examples fine at FS, but after making the appropriate pin config changes and tinyUSB config changes, I cannot seem to get any of the HS examples to work. They compile and flash fine but when plugging the board in, nothing happens at all. I've tested the hardware by compiling a HID example with the ST USB library and it all worked fine. I just can't figure out why the tinyUSB HS examples aren't working and without any expensive HS USB analyser, it seems like it may be a difficult task to debug. In addition, I've tried flashing a NUCLEO-H743ZI2 with a HS example without success, though I don't entirely trust the connection from the dev board to USB PHY was great using dupoint breadboard jumper wires.

I'll keep digging and debugging, but any help I can get here might save me a heap of time. Where is the most obvious place I should be looking for issues? Any help is greatly appreciated.

Cheers,
Ben

**Comment by hathach** [comment, marked as answer]
make sure you run with stock example and add your own bsp based on other similar supported bsp such as h743eval https://github.com/hathach/tinyusb/blob/master/hw/bsp/stm32h7/boards/stm32h743eval/board.mk you need to run make with `PORT=1` as option (h743eval has it set to 1 by default).  Always start with stock example first.

**Reply by BennyEvans** [reply]
Hi @hathach , I've just created my own bsp, using the h743eval for reference but updating clocks, makefiles, pin config etc. for this board. I'm compiling the stock examples without modification. It all compiles fine but again, when pluggin in via USB, it doesn't seem to be doing anything. The flashing light configured in the cdc_msc example flashes at a consistent rate when plugging in and out the USB cable. While playing around with a project in STMcubeIDE, I did note the the USB HS interrupt either never fired, or fired once occationally when plugging in the USB cable. I didn't debug much further at that time but might look at that further over the next few days. Does this provide any more hints as to what the issue could be?

**Reply by hathach** [reply]
last time I checked, it works fine with 743eval. Therefore it isn't an issue with the software, you should check with h743 eval schematics and yours, maybe try to run an demo example that comes with your boards to verify the hardware work first. Note: H743 requires external PHY to work with USB HS , on h743eval it is USB3320

**Reply by BennyEvans** [reply]
I have run a USB HS demo with the hardware and it does work, so I don't believe it to be the hardware. I'm using a USB3300 PHY (https://www.waveshare.com/wiki/USB3300_USB_HS_Board).

**Reply by hathach** [reply]
It could be clock or pin set up, it is a good idea to do a full check and compare against ST H743eval board for the bsp definition. Also, please try to compile and run with `make BOARD=yourboard LOG=3`, you can either get the log via uart (default) or segger RTT which could give some hint on the issue https://github.com/hathach/tinyusb/blob/master/docs/getting_started.md . Since I don't have your hardware, you will need to troubleshoot the issue, any log or info update would help for analysis and suggestion.

**Reply by BennyEvans** [reply]
I've done a bit more investigation but still haven't found a solution. The clock configuration seems correct as I can successfully run the FS USB examples without the external PHY. I've taken the HS pin configuration from a working dev board example using the ST USB libraries so I don't believe that's the culprit either.

When compiling for HS with LOG=3, I only recieve the following log output when running the example:
> USBD init
> CDC init
> MSC init

I can plug the USB cable in/out and no other log messages are written (the application is still running though and has not locked up). Interestingly, the OTG_HS interrupt (using OTG_HS_IRQHandler) is only ever triggered at initialisation too. Plugging in/out the USB cable does nothing. The value of the Core Interrupt Register (GINTSTS) when the first and only USB interrupt fires is 0x04008c20. For reference I've attached the register description and the register itself below.

![image](https://user-images.githubusercontent.com/2138027/125411789-99b7ec80-e401-11eb-9d44-281ce9ae54d9.png)

![image](https://user-images.githubusercontent.com/2138027/125411550-5cebf580-e401-11eb-8504-30477b9378fd.png)

USB suspend and Early suspend are set here but I'm not yet at the stage in knowing if this is correct or not.

Another difference between my board and the existing H7 supported boards is that I'm using a Silicon Revision V chip. There were some changes to USB in this revision but ST claim that it should be software compatible. If someone is using a Rev. V STM32H7 and succsessfully running a HS example, that may also rule out that as a potential problem.

Do you have any opinion on where I should look next to provide anymore information?

**Reply by hathach** [reply]
Unfortunately, I don't have any particular suggestion to you, since I don't have any further experience with your board and new rev mcu. I also had an really hard time getting this running on the H743 eval board myself. The ST OTG configuration is quite complicated and each variant seems to have a small but critical difference to others. However, IMO, the new rev shouldn't be the issue. It is probably still due to pin/clock config, maybe you should post on the ST Community forum, there are more people doing more work with the ST MCU. 

Looking at the dcd initialization https://github.com/hathach/tinyusb/blob/master/src/portable/st/synopsys/dcd_synopsys.c#L429 can be useful, also please confirm
- If the USB IRQ is enabled
- Verify pin, bsp
- Compare major register OTG, PHY, INT, Control etc... is relative the same when using with both tinyusb and st-stack

**Reply by BennyEvans** [reply]
While debugging this I managed to get it in a working state. I still haven't found the issue and there is still some strange phenomenon, but I can get it into a working state by adding some delays before and after the call to tusb_init(). I also added a small delay in dcd_connect after reading the datasheet but this does not seem to have an effect.

Interestingly, after the initial enumeration succeeds, I can reduce the delays without issue. But after powering down the chip for 10-15 seconds and powering back up, I need the longer delays to get everything going again. Additionally, the length of the delays seem to illogically affect the CDC throughput and I also somehow manage to get more throughput when CFG_TUSB_DEBUG = 1 over when CFG_TUSB_DEBUG = 0. Doesn't make sense....

As a sanity check, I've purchased an additional USB3300 PHY to try as something doesn't seem right from the results I'm getting. I'll keep you updated with the results on this front.

I'm also seeing an issue with the CDC echo when sending exactly 256 bytes. Other lengths work fine, but when echoing a 256 byte string, the port seems to lockup. I haven't performed much digging on this issue yet as I've been trying to solve my other issue but once I have some detail to provide for this, I'll create another discussion.

**Reply by hathach** [reply]
Thanks for the update, maybe the PHY need a extra delay for power stable or something. It is beyond my knowledge, hope that you could figure some busy status bit to wait rather than blindly delay which could be off sometime. And thank you for opening the cdc with hs, it is spotted on.

**Reply by BennyEvans** [reply]
Final update for this discussion before marking it resolved. This update is mainly for others with a Waveshare STM32H7 board and USB3300 PHY.

I received a new PHY, however, it made no difference. I decided I'd rework the board to connect the PHY reset pin to a GPIO as this waveshare board has it NC, although the PHY board does have it tied to GND. I connected it to PD14 and modified the code to init PD14, bring it high for 5ms (with a delay) and then set it back to low for 50ms before calling `tusb_init()`. This seems to get everything up and running almost every time, if not every time. I've also had a python scipt running for a couple of hours, continually writing various lengths of data to the CDC serial port with various delays and comparing the received data with what was sent. It has run without problem so everything seems to be stable. I still somehow get different throughput by varying the initial delay, but currently I'm putting that down to a weird hardware issue. Without the board design files, it's hard to see if the ULPI hardware design guide was followed when making this board but from what I can see, it doesn't look like trace length matching or impedance control was even a thought. If you are reading this and are considering the waveshare board for HS USB development, I'd advise against it, although currently it doesn't seem like there are many cheap alternatives without creating your own board.

Additionally, if you're still having problems, it seems ST have some undocumented registers on their F7 and H7 chips (and maybe more). If you download the STMCube F7 firmware, you'll find two functions named `USB_ULPI_Read` and `USB_ULPI_Write` in `stm32f7xx_lp_modes.c`. You can use these functions to read/write the ULPI PHY registers which can assist in debugging issues with the PHY coming up. It's a bit frustrating that these registers aren't documented by ST but I can confirm that they do work on the H7 series.

I was considering creating a PR with a bsp for this waveshare board, but considering a small rework is required to get everything to work smoothly, and the strange throughput issue still exists, I don't think it's a good idea **yet**.

Thanks again for your help.

**Reply by HiFiPhile** [reply]
It's interesting to know that you are also having problem with a Waveshare board...

Initially I was developing LPC4357 with Waveshare Open4357 and it was a nightmare, I've never see a dev board with such bad signal integrity. Core module is 4-layer PCB without GND layer and base board is only 2 layer, LCD flicks when accessing SDRAM.
In the end switched to a LPCXpresso.

**Reply by hathach** [reply]
@BennyEvans thank you very much for the update, I am sure this help lots of others use that are having the same board. Too bad that you have to rework to get the waveshare H743 to work. Though their hardware is not perfect, the cost is only 1/3 of  ST H743 eval. Do you think it is still possible that we could put some initial phy init (copied from part of the dcd_init() ) + extra delay as you mentioned in the board_init(). If we could get it running as proof of concept, it may be still valuable to others. We could put your USB HS PHY note on the bsp source file to notify others as well.

**Reply by BennyEvans** [reply]
@hathach I'll still put together a bsp for the board with a lengthy paragraph at the top of the source file noting the rework, jumper settings etc. It'll probably be a week or so though before I submit a PR as I'll need to clean it all up when I get the chance.

As waveshare create different boards for different series/manufacturers, I'll probably call the bsp "waveshare_openh743i". Is this ok or would you prefer a different structure?

**Reply by hathach** [reply]
the name is perfect, take your time, there is no rush, it is purely to help other with the same board.

**Reply by BennyEvans** [reply]
Bit of an update on the BSP package for this board. I'm happy to submit a PR for the board with FS working, however, I cannot get HS to work reliably. I'm sure if I commit in it's current state, we'll end up with more questions about why it doesn't work. I can commit the code as is with a disclaimer that HS isn't working fully, but want your opinion on what I should do here.

I'm still digging to find why it isn't working reliably.

**Reply by BennyEvans** [reply]
My custom HS CDC only code in STMCube seems to work very reliably now, but when compiling the cdc_msc example with the GNU Arm Embedded Toolchain, I can't get the PHY running at all. I have a loop at startup that continually attempts to read from the PHY, resetting the PHY on a timeout but it just never responds. I'll keep digging and might probe some signals with an oscilloscope to see what's happening. But it might take me longer than expected to get a HS BSP.

The FS examples work fine in either toolchain and in every example I've tested so far.

**Reply by BennyEvans** [reply]
Apologies for my rambling, it's been driving me crazy but both HS and FS seem to be working fairly reliably with my BSP in both toolchains now. The addition of `__HAL_RCC_SYSCFG_CLK_ENABLE()` also helped :confounded:

Expect a PR within the next couple of days after I've tested some more.

**Reply by hathach** [reply]
no problems at all, I appreciate your effort and update on the progress. I am glad you figured it out. It is really hard to work with HS PHY from my last experience getting it running :)

**Reply by BennyEvans** [reply]
I've just submitted the PR. It all seems to be working well. Cheers.
